Documentation/internals/security-identities.rst

```diff
@@ -0,0 +1,64 @@
+.. only:: not (epub or latex or html)
+
+    WARNING: You are looking at unreleased Cilium documentation.
+    Please use the official rendered version released here:
+    https://docs.cilium.io
+
+.. _security_identities:
+
+*******************
+Security Identities
+*******************
+
+Security identities are generated from labels. They are stored as ``uint32``,
+which means the maximum limit for a security identity is ``2^32 - 1``. The
+minimum security identity is ``1``, due to the reserved identity range being
+from ``1`` to ``255``. See the `internal code documentation
+<https://pkg.go.dev/github.com/cilium/cilium/pkg/identity#NumericIdentity>`_.
+
+.. note::
+
+   Identitiy 0 is not a valid value. If it shows up in Hubble output, this
+   means the identity was not found. In the eBPF datapath, it has a special
+   role where it denotes "any identity", i.e. as a wildcard allow in policy
+   map.
+
+There are two caveats to the above however:
+
+1) ClusterMesh
+2) Identities generated from CIDR-based policies
+
+For (1), an extra 8 bits are added as the ``cluster-id`` which identifies the
+cluster in the ClusterMesh, into the 3rd octet as shown by ``0x00FF0000``. This
+makes the security identity's upper limit ``2^24 - 1``. This does not apply to
+CIDR identities however, see (2).
+
+For (2), CIDR identities are local to each node. This means that they're not
+subject to the global limit of security identities (``2^16 - 1``). CIDR
+identities begin from ``1`` and end at ``16777215``, however since they're
+shifted by ``24``, this makes their effective range ``1 | (1 << 24)`` to
+``16777215 | (1 << 24)`` or from ``16777217`` to ``33554431``.  The reason that
+CIDR identities are local is because of the limitation imposed by the VXLAN or
+Geneve virtual network field size being 24 bits.
```

Comment:
```suggestion
``16777215 | (1 << 24)`` or from ``16777217`` to ``33554431``.  These identities do not
need to fit inside a VXLAN or Geneve virtual network field and hence are not limited to 24 bits.
```

Reply:
^^ I guess if we were being super clear then maybe we would just call out earlier as a separate point that there are certain constraints on the identity space, primarily that any identity which needs to be shared between nodes must fit in the packet header fields that we use to encode the identity. This means fitting inside the 24-bit VNI (virtual network identifiier) field in VXLAN or Geneve headers. Then we don't need to repeat these constrains as part of each of these later explanations. But :shrug: this documentation is already greatly improving the status quo around explanation of these crucial internals, we don't need to make the docs perfect the first time.

Reply:
Yeah, I opted to remove this sentence all together and just include the explanation once from https://github.com/cilium/cilium/pull/16716#discussion_r779877262.